### PR TITLE
Jinja2: add posix path filter

### PR DESCRIPTION
**Pull request by johnwparent**
Apply filter to bootstrap template, without it, Windows paths are written with literal `\` and the resulting yaml becomes un-processable due to "invalid escape sequences"

### CODE REVIEW

File: share/spack/templates/bootstrap/spack.yaml [hunk 1]
```diff
@@ -10,18 +10,18 @@ spack:
...
-      root: {{ store_path }}
+      root: {{ store_path|posixpaths }}
 
   packages:
     python:
       buildable: false
       externals:
         - spec: "{{ python_spec }}"
-          prefix: "{{ python_prefix }}"
+          prefix: "{{ python_prefix|posixpaths }}"
```
Comment (alalazo) [inline share/spack/templates/bootstrap/spack.yaml:24]: I'm confused. [The documentation](https://spack.readthedocs.io/en/latest/getting_started.html#step-2-install-and-setup-spack) seems to suggest we are using windows paths , when on Windows?  :thinking:

Comment (johnwparent) [reply share/spack/templates/bootstrap/spack.yaml:24]: Sorry for any confusion. Windows paths should, in general, only be used when we're calling out to the Windows filesystem and basically nowhere else, as improperly escaped backslashes can wreak havoc, as is the case in the bug this PR fixes, and it's easy to miss/forget, particularly when some operations are processing the escapes into single backslashes and those backslashes are then being evaluated literally. 

In general, we should be taking the approach of any path coming _into_ spack should be made posix, and only when calling to the Windows filesystem should a path be converted to its platform dependent scheme. 

I'm not sure why you're pointing to the getting started guide for Windows. Yes, we support Windows as a platform, but that does not be we should blindly be using Windows paths everywhere, the issue is far more nuanced than "this is the path structure for this platform always."
I also don't see where that doc says anything about "we are using Windows paths" it just says we support Windows and the path to the Spack cmd entrypoint should be a backslash because it will be entered into Windows CMD which only understands backslashes as it is a DOS holdover.

Comment (johnwparent) [reply share/spack/templates/bootstrap/spack.yaml:24]: FWIW without this PR, much of the bootstrapping command is broken on Windows, with only the very basic behavior functioning, and the rest erroring out with an "improper escape sequence" when the bootstrapping `spack.yaml` is read by Spack.

Comment (alalazo) [reply share/spack/templates/bootstrap/spack.yaml:24]: > I'm not sure why you're pointing to the getting started guide for Windows.

Because of this example:
```yaml
packages:
  cmake:
    externals:
    - spec: cmake@3.19
      prefix: 'c:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake'
    buildable: False
  ninja:
    externals:
    - spec: ninja@1.8.2
      prefix: 'c:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja'
    buildable: False
```
that seems to suggest we use native path on Windows. Are they currently both allowed?

Comment (alalazo) [reply share/spack/templates/bootstrap/spack.yaml:24]: > In general, we should be taking the approach of any path coming into spack should be made posix, and only when calling to the Windows filesystem should a path be converted to its platform dependent scheme.

Here we are not doing that, right? We are just having an ad-hoc change for a specific template - which can be out of sync with other parts of Spack. Or am I missing something? Wouldn't this lead to having some paths be native and some be posix in config files when on Windows?

Comment (johnwparent) [reply share/spack/templates/bootstrap/spack.yaml:24]: Ok I was distracted by the pin to another section, thanks for clarifying. 

Right now spack is in a bit of a weird state where most things are just using the `os.path` module whereas others are using pathlib and other still are just hardcoding posix style paths. This has led to a bit of an intractable scenario where I am forced to chase down niche cases in a large codebase or perform massive refactors just to get things consistent enough to fix a small bug (like this case, where a simple, small fix does the job, but actually fixing the problem at a conceptual scope is a massive effort as it would require touching vast portions of Spack). This fix (and the many others like it) are stop gaps, not intended to live for very long, but are needed to make things functional on Windows while we chip away at the larger refactors needed to actually solve the underlying problem that, as you point out, this PR does not. 

As I stated in my earlier comment, some parts of Spack _should_ have posix paths only where others _should_ have native paths, depending on the context. This is not a "one or the other". 

This PR is an example of a place where we should _not_ have native paths regardless of what the rest of Spack is doing.. The above example (and the rest of the spack.yaml files) is one I actually debated refactoring in this PR as well, as it should also be a posix path (Spack is the only thing reading it and it's not being piped directly to sys calls, which are my criteria for "should this be native") and I'd imagine a path like `C:\Users` would be dangerous. However, as you point out the changes are largely ad-hoc, I'd generally like to reduce that, and that hasn't broken anything _yet_. If it does, I'll apply the same ad-hoc solution until we get the actual refactors in. 


This is a bug in Spack that is broken right now for users on Windows. A new user should be able to actually _use_ spack without hitting basic bugs like this that make certain features (like manual bootstrapping) completely unusable. The ideal solution (one that we are working on in addition to fixing bugs like this and trying to achieve parity with posix systems) is a complete remodel of how Spack does paths. However, that's a ways off given the current development and review velocity, and this needs a fix now. Realistically, all Spack paths should be pathlib `PurePosixPaths` until they need to interact with an actual concrete filesystem. Until we have that, this fix is needed.

In short. Yes, this fix is ad-hoc and is necessary to correct an already intractably convoluted state re: paths that Spack is in, but is by no means intended to be ideal or even last more than a few months.

Comment (alalazo) [reply share/spack/templates/bootstrap/spack.yaml:24]: > Right now spack is in a bit of a weird state where most things are just using the os.path module whereas others are using pathlib and other still are just hardcoding posix style paths.

Would issues be solved on the Windows side if we used `pathlib` only? To be clear, I'm not much concerned about this specific PR (I'm just commenting, not requesting anything). Rather, I'd like to understand if we have a strategy to consolidate in fewer places these platform differences, that many times increase the complexity of simple operations. 

> This fix (and the many others like it) are stop gaps, not intended to live for very long, but are needed to make things functional on Windows while we chip away at the larger refactors needed to actually solve the underlying problem that, as you point out, this PR does not.

Are we marking these stop gaps in any way, so they'll be easy to find when we come to a refactor?

Comment (johnwparent) [reply share/spack/templates/bootstrap/spack.yaml:24]: > Would issues be solved on the Windows side if we used `pathlib` only? To be clear, I'm not much concerned about this specific PR (I'm just commenting, not requesting anything). Rather, I'd like to understand if we have a strategy to consolidate in fewer places these platform differences, that many times increase the complexity of simple operations.

Yes! I often see you introducing pathlib into the codebase and I am eternally grateful to you for doing that, it makes my life 10000% easier and Windows support that much more robust. It won't eliminate 100% of the issues, but it would make any corner cases much more robust and trivial. In addition to other active efforts, @markus-ferrell and I have been refactoring large swathes of the codebase to use pathlib (I have a pr refactoring all of llnl/util and markus has two that refactor all of spack/util) based of a branch I have that refactors the entire codebase but that I was unable to maintain due to merge conflicts that took longer to resolve than I was able to develop.

> > This fix (and the many others like it) are stop gaps, not intended to live for very long, but are needed to make things functional on Windows while we chip away at the larger refactors needed to actually solve the underlying problem that, as you point out, this PR does not.
> 
> Are we marking these stop gaps in any way, so they'll be easy to find when we come to a refactor?

Generally I leave a todo for my GH username, and I have a local script that searches for TODOs assigned to me and spits out a "Spack to do list" every morning, which, I forgot to do here, so thanks for pointing that out.

Comment (johnwparent) [reply share/spack/templates/bootstrap/spack.yaml:24]: > To be clear, I'm not much concerned about this specific PR (I'm just commenting, not requesting anything). Rather, I'd like to understand if we have a strategy to consolidate in fewer places these platform differences, that many times increase the complexity of simple operations.

Thank you for clarifying, it initially seemed that you were in favor of rejecting this PR flat out due to the larger considerations, hence my defense of this PR. I now see I misunderstood, so thank you again for making that clear.

**Comment by johnwparent** [comment]
Closing, there are better ways to do this.
